src/MonitorProfiler/ProbeInstrumentation/ProbeInstrumentation.cpp

```diff
@@ -0,0 +1,300 @@
+    m_pCorProfilerInfo(profilerInfo),
+    m_pLogger(logger),
+    m_probeFunctionId(0),
+    m_pAssemblyProbePrep(nullptr)
+{
+}
+
+HRESULT ProbeInstrumentation::RegisterFunctionProbe(FunctionID enterProbeId)
+{
+    lock_guard<mutex> lock(m_probePinningMutex);
+
+    if (HasPinnedProbe())
+    {
+        // Probes have already been pinned.
+        return E_FAIL;
+    }
+
+    m_pLogger->Log(LogLevel::Debug, _LS("Received probes"));
+
+    m_pAssemblyProbePrep.reset(new (nothrow) AssemblyProbePrep(m_pCorProfilerInfo, enterProbeId));
+    IfNullRet(m_pAssemblyProbePrep);
+
+    // JSFIX: Validate the probe's signature before pinning it.
+    m_probeFunctionId = enterProbeId;
+
+    return S_OK;
+}
+
+HRESULT ProbeInstrumentation::InitBackgroundService()
+{
+    m_probeManagementThread = thread(&ProbeInstrumentation::WorkerThread, this);
+    return S_OK;
+}
+
+void ProbeInstrumentation::WorkerThread()
+{
+    HRESULT hr = m_pCorProfilerInfo->InitializeCurrentThread();
+    if (FAILED(hr))
+    {
+        m_pLogger->Log(LogLevel::Error, _LS("Unable to initialize thread: 0x%08x"), hr);
+        return;
+    }
+
+    while (true)
+    {
+        PROBE_WORKER_PAYLOAD payload;
+        hr = m_probeManagementQueue.BlockingDequeue(payload);
+        if (hr != S_OK)
+        {
+            break;
+        }
+
+        switch (payload.instruction)
+        {
+        case ProbeWorkerInstruction::INSTALL_PROBES:
+            hr = InstallProbes(payload.requests);
+            if (hr != S_OK)
+            {
+                m_pLogger->Log(LogLevel::Error, _LS("Failed to install probes: 0x%08x"), hr);
+            }
+            break;
+
+        case ProbeWorkerInstruction::UNINSTALL_PROBES:
+            hr = UninstallProbes();
+            if (hr != S_OK)
+            {
+                m_pLogger->Log(LogLevel::Error, _LS("Failed to uninstall probes: 0x%08x"), hr);
+            }
+            break;
+
+        default:
+            m_pLogger->Log(LogLevel::Error, _LS("Unknown message"));
+            break;
+        }
+    }
+}
+
+void ProbeInstrumentation::ShutdownBackgroundService()
+{
+    m_probeManagementQueue.Complete();
+    m_probeManagementThread.join();
+}
+
+HRESULT ProbeInstrumentation::RequestFunctionProbeInstallation(
+    ULONG64 functionIds[],
+    ULONG32 count,
+    ULONG32 argumentBoxingTypes[],
+    ULONG32 argumentCounts[])
+{
+    m_pLogger->Log(LogLevel::Debug, _LS("Probe installation requested"));
+
+    if (!HasPinnedProbe())
+    {
+        return S_FALSE;
+    }
+
+    vector<UNPROCESSED_INSTRUMENTATION_REQUEST> requests;
+    requests.reserve(count);
+
+    ULONG32 offset = 0;
+    for (ULONG32 i = 0; i < count; i++)
+    {
+        vector<ULONG32> tokens;
+        tokens.reserve(argumentCounts[i]);
+        ULONG32 j;
+
+        for (j = 0; j < argumentCounts[i]; j++)
+        {
+            tokens.push_back(argumentBoxingTypes[offset+j]);
```

Comment:
Should the `E_INVALIDARG` check be above this? Meaning, if `offset + j` really is greater than UINT32_MAX, we wouldn't want to try indexing into `argumentBoxingTypes` with it?